### PR TITLE
⚡️ MerkleProof Verify Calldata Version

### PR DESCRIPTION
**Pull request by Vectorized**
## Description

This is for #242. 

After some marinating, I've decided to change verify to take in a calldata array instead of a memory array.

Since MerkleProof is used for saving on-chain storage afterall,  
the data is almost certainly likely to be from an external call.

Since Solmate is an opinionated minimalist repo, I'd recommend just including the version that is most likely to be relevant.

Also, sorry for missing the empty trailing space in the previous PR.

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [X] Ran `forge snapshot`?
- [X] Ran `npm run lint`?
- [X] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

## Gas Benchmarks

The gas snapshots will have higher gas because of the call overhead required to test calldata.  

Once you factor in the call overhead, these are the numbers:

```
[PASS] testValidProofSupplied() (gas: 2464)
[PASS] testValidProofSuppliedCalldata() (gas: 2280)
[PASS] testVerifyEmptyMerkleProofSuppliedLeafAndRootDifferent() (gas: 1733)
[PASS] testVerifyEmptyMerkleProofSuppliedLeafAndRootDifferentCalldata() (gas: 1565)
[PASS] testVerifyEmptyMerkleProofSuppliedLeafAndRootSame() (gas: 1727)
[PASS] testVerifyEmptyMerkleProofSuppliedLeafAndRootSameCalldata() (gas: 1537)
[PASS] testVerifyInvalidProofSupplied() (gas: 2460)
[PASS] testVerifyInvalidProofSuppliedCalldata() (gas: 2322)
```

### CODE REVIEW

File: src/utils/MerkleProof.sol [hunk 1]
```diff
@@ -37,4 +37,27 @@ library MerkleProof {
...
+            } {
+                let loadedData := calldataload(data)
+                let scratch := shl(5, gt(computedHash, loadedData))
+                mstore(scratch, computedHash)
+                mstore(xor(scratch, 0x20), loadedData)
+                computedHash := keccak256(0x00, 0x40)
+            }
+            isValid := eq(computedHash, root)
+        }
+    }
```
Comment (transmissions11) [inline src/utils/MerkleProof.sol:62]: mind adding some comments here?

File: src/utils/MerkleProof.sol [hunk 2]
```diff
@@ -37,4 +37,27 @@ library MerkleProof {
             isValid := eq(computedHash, root)
         }
     }
-}
\ No newline at end of file
+
+    function verifyCalldata(
```
Comment (transmissions11) [inline src/utils/MerkleProof.sol:41]: imo lets rename this to verify and the other to verifyMemory

File: src/test/MerkleProof.t.sol [hunk 3]
```diff
@@ -37,8 +37,56 @@ contract MerkleProofTest is DSTestPlus {
...
+        return MerkleProof.verify(proof, root, leaf);
+    }
+
+    function verifyCalldata(
+        bytes32[] calldata proof,
+        bytes32 root,
+        bytes32 leaf
+    ) external pure returns (bool) {
+        return MerkleProof.verifyCalldata(proof, root, leaf);
+    }
```
Comment (transmissions11) [inline src/test/MerkleProof.t.sol:91]: these feel kinda pointless unless we make any assertions?

File: src/utils/MerkleProof.sol (hunk 2 above)
Comment (transmissions11) [reply src/utils/MerkleProof.sol:41]: or actually can we just remove verifyMemory?

Comment (Vectorized) [reply src/utils/MerkleProof.sol:41]: Let's remove the memory version.

**Comment by Vectorized** [comment]
Ok, remove the memory version.   
This will also indirectly enforce good practice.

Since Solmate is an opinionated minimalist repo, it makes more sense to have just one recommended version.

If someone somehow needs the memory version, they are always free to copypasta it and modify.

**Comment by transmissions11** [comment]
amazing work, lgtm!

**Pull request merged**